extra_rpaths in compilers.yaml not working

In #4256, I added a package for OpenBabel. After installation, I ran `ldd -r` on the libraries and found that they were linking to the wrong `libstdc++` library:
```
$ ldd -r libopenbabel.so
libopenbabel.so: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by libopenbabel.so)
...
libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00002b47e587d000)
```
By adding the GCC compiler to my `LD_LIBRARY_PATH`, I was able to resolve this problem. I tried making the following change in `compilers.yaml` to do this automatically:
```yaml
- compiler:
    environment: {}
    extra_rpaths:
    - /soft/spack/opt/spack/linux-x86_64/gcc-5.3.0/gcc-6.1.0-q2zosj3igepi3pjnqt74bwazmptr5gpj/lib64
    flags: {}
    modules: []
    operating_system: centos6
    paths:
      cc: /soft/spack/opt/spack/linux-x86_64/gcc-5.3.0/gcc-6.1.0-q2zosj3igepi3pjnqt74bwazmptr5gpj/bin/gcc
      cxx: /soft/spack/opt/spack/linux-x86_64/gcc-5.3.0/gcc-6.1.0-q2zosj3igepi3pjnqt74bwazmptr5gpj/bin/g++
      f77: /soft/spack/opt/spack/linux-x86_64/gcc-5.3.0/gcc-6.1.0-q2zosj3igepi3pjnqt74bwazmptr5gpj/bin/gfortran
      fc: /soft/spack/opt/spack/linux-x86_64/gcc-5.3.0/gcc-6.1.0-q2zosj3igepi3pjnqt74bwazmptr5gpj/bin/gfortran
    spec: gcc@6.1.0
    target: x86_64
```
but it appears that `extra_rpaths` isn't working. I reinstalled Open Babel but it still links to the system `libstdc++` instead of the one in `$GCC_ROOT/lib64`. Adding it manually to `LD_LIBRARY_PATH` works:
```
$ ldd -r /blues/gpfs/home/ajstewart/spack/opt/spack/linux-centos6-x86_64/gcc-6.1.0/openbabel-2.4.1-xqe2leslpleij3g7u4dpje7joxcdydjh/lib/libopenbabel.so
...
	libstdc++.so.6 => /soft/spack/opt/spack/linux-x86_64/gcc-5.3.0/gcc-6.1.0-q2zosj3igepi3pjnqt74bwazmptr5gpj/lib64/libstdc++.so.6 (0x00002b4b18855000)
```
@scheibelp?

as commented in https://github.com/spack/spack/issues/10308#issuecomment-716391950 I think the linker itself already needs `libimf.so` to run (note how it says "error while loading shared libraries" and not "cannot find -limf" which is the usual error when an object on the linker line can't be found), while the extra rpaths are taken into account only for the object to link, hence the documented way to use a `modules` file instead should ensure that the required path also ends up in the `LD_LIBRARY_PATH`

--- a/lib/spack/spack/build_systems/intel.py
+++ b/lib/spack/spack/build_systems/intel.py
@@ -1360,7 +1360,8 @@ class IntelPackage(PackageBase):
             # in compiler releases, then we need to search for libimf.so instead
             # of this static path.
             for lib in LLVMgold_libs:
-                p = subprocess.Popen(['patchelf', '--print-rpath', lib], stdout=subprocess.PIPE)
+                p = subprocess.Popen(['patchelf', '--print-rpath', lib],
+                                     stdout=subprocess.PIPE)
                 rpath = ':'.join([str(p.communicate()[0].decode()).strip(),
                                   '$ORIGIN/../compiler/lib/intel64_lin'])
                 subprocess.call(['patchelf', '--set-rpath', rpath, lib])

--- a/lib/spack/spack/build_systems/intel.py
+++ b/lib/spack/spack/build_systems/intel.py
@@ -8,6 +8,7 @@ import glob
 import inspect
 import os
 import re
+import subprocess
 import sys
 import tempfile
 import xml.etree.ElementTree as ElementTree
@@ -1331,6 +1332,36 @@ class IntelPackage(PackageBase):
 
         debug_print(os.getcwd())
         return
+
+    @property
+    def base_lib_dir(self):
+        """Provide the library directory located in the base of Intel installation.
+        """
+        d = self.normalize_path('')
+        d = os.path.join(d, 'lib')
+
+        debug_print(d)
+        return d
+
+    @run_after('install')
+    def modify_LLVMgold_rpath(self):
+        """Add libimf.so and other required libraries to the RUNPATH of LLVMgold.so.
+
+        These are needed explicitly at dependent link time when
+        `ld -plugin LLVMgold.so` is called by the compiler.
+        """
+        if self._has_compilers:
+            LLVMgold_libs = find_libraries('LLVMgold', self.base_lib_dir,
+                                           shared=True, recursive=True)
+            # Ignore ia32 entries as they mostly ignore throughout the rest
+            # of the file.
+            # The first entry in rpath preserves the original, the seconds entry
+            # is the location of libimf.so. If this relative location is changed
+            # in compiler releases, then we need to search for libimf.so instead
+            # of this static path.
+            rpath = "$ORIGIN/../lib:$ORIGIN/../compiler/lib/intel64_lin"
+            for lib in LLVMgold_libs:
+                subprocess.call(['patchelf', '--set-rpath', rpath, lib])
 
     # Check that self.prefix is there after installation
     run_after('install')(PackageBase.sanity_check_prefix)

--- a/lib/spack/spack/build_systems/intel.py
+++ b/lib/spack/spack/build_systems/intel.py
@@ -8,7 +8,6 @@ import glob
 import inspect
 import os
 import re
-import subprocess
 import sys
 import tempfile
 import xml.etree.ElementTree as ElementTree
@@ -25,6 +24,7 @@ from llnl.util.filesystem import (
     install,
 )
 
+import spack.error
 from spack.build_environment import dso_suffix
 from spack.package import InstallError, PackageBase, run_after
 from spack.util.environment import EnvironmentModifications
@@ -1360,11 +1360,14 @@ class IntelPackage(PackageBase):
             # in compiler releases, then we need to search for libimf.so instead
             # of this static path.
             for lib in LLVMgold_libs:
-                p = subprocess.Popen(['patchelf', '--print-rpath', lib],
-                                     stdout=subprocess.PIPE)
-                rpath = ':'.join([str(p.communicate()[0].decode()).strip(),
+                if not self.spec.satisfies('^patchelf'):
+                    raise spack.error.SpackError(
+                        'Attempting to patch RPATH in LLVMgold.so.
+                        `patchelf` dependency should be set in package.py')
+                patchelf = Executable('patchelf')
+                rpath = ':'.join([patchelf('--print-rpath', lib, output=str).strip(),
                                   '$ORIGIN/../compiler/lib/intel64_lin'])
-                subprocess.call(['patchelf', '--set-rpath', rpath, lib])
+                patchelf('--set-rpath', rpath, lib)
 
     # Check that self.prefix is there after installation
     run_after('install')(PackageBase.sanity_check_prefix)

--- a/lib/spack/spack/build_systems/intel.py
+++ b/lib/spack/spack/build_systems/intel.py
@@ -1362,8 +1362,9 @@ class IntelPackage(PackageBase):
             for lib in LLVMgold_libs:
                 if not self.spec.satisfies('^patchelf'):
                     raise spack.error.SpackError(
-                        'Attempting to patch RPATH in LLVMgold.so.
-                        `patchelf` dependency should be set in package.py')
+                        'Attempting to patch RPATH in LLVMgold.so.'
+                        + '`patchelf` dependency should be set in package.py'
+                    )
                 patchelf = Executable('patchelf')
                 rpath = ':'.join([patchelf('--print-rpath', lib, output=str).strip(),
                                   '$ORIGIN/../compiler/lib/intel64_lin'])

--- a/lib/spack/spack/build_systems/intel.py
+++ b/lib/spack/spack/build_systems/intel.py
@@ -1359,8 +1359,10 @@ class IntelPackage(PackageBase):
             # is the location of libimf.so. If this relative location is changed
             # in compiler releases, then we need to search for libimf.so instead
             # of this static path.
-            rpath = "$ORIGIN/../lib:$ORIGIN/../compiler/lib/intel64_lin"
             for lib in LLVMgold_libs:
+                p = subprocess.Popen(['patchelf', '--print-rpath', lib], stdout=subprocess.PIPE)
+                rpath = ':'.join([str(p.communicate()[0].decode()).strip(),
+                                  '$ORIGIN/../compiler/lib/intel64_lin'])
                 subprocess.call(['patchelf', '--set-rpath', rpath, lib])
 
     # Check that self.prefix is there after installation

--- a/var/spack/repos/builtin/packages/intel-parallel-studio/package.py
+++ b/var/spack/repos/builtin/packages/intel-parallel-studio/package.py
@@ -13,6 +13,8 @@ class IntelParallelStudio(IntelPackage):
 
     maintainers = ['rscohn2', 'danvev']
 
+    depends_on('patchelf', type='build')
+
     # As of 2016, the product comes in three "editions" that vary by scope.
     #
     # In Spack, select the edition via the version number in the spec, e.g.:

--- a/var/spack/repos/builtin/packages/intel/package.py
+++ b/var/spack/repos/builtin/packages/intel/package.py
@@ -14,6 +14,8 @@ class Intel(IntelPackage):
 
     maintainers = ['rscohn2', 'danvev']
 
+    depends_on('patchelf', type='build')
+
     # Same as in ../intel-parallel-studio/package.py, Composer Edition,
     # but the version numbering in Spack differs.
     version('20.0.4',              sha256='ac1efeff608a8c3a416e6dfe20364061e8abf62d35fbaacdffe3fc9676fc1aa3', url='https://registrationcenter-download.intel.com/akdlm/irc_nas/tec/17117/parallel_studio_xe_2020_update4_composer_edition.tgz')

